src/charm.py

```diff
@@ -258,8 +286,9 @@ def _on_config_changed(self, event: EventBase) -> None:
         if zk_jaas_changed or properties_changed:
             self.on[f"{self.restart.name}"].acquire_lock.emit()
 
-        # update client_properties whenever possible
+        # update these whenever possible
         self.config_manager.set_client_properties()
+        self.update_external_services()
```

Comment:
Note that `charm.py` isn't shared between K8s/VM, they're unique for both (pebble-ready differences as an example) so this is fine to have here.

Reply:
Although I haven't tested it against VM yet, I am fairly confident that it'll copy over as is with minimal changes, all the core logic for VM was retained.

Reply:
Yes, definitely. `charm.py` can be substrate specific. However, right now we have a bit of common logic and copy and paste there too. Actually, I was recently discussing this with Raul as well, as suggesting to try to separate and factor out from `charm.py` the "common" part, such that `charm.py` only has "substrate" specific stuff. 

Note that I mentioned this here since it was within a sort-of "general" hook, the _changed_config one. The substrate specific logic in substrate specific hooks (e.g. pebble ready event) can surely stay!

Reply:
Yeah, I'd like to work on leaning-in to the copied code across both, in `charm.py` and in the integration-tests especially. I'll keep that in mind, but probably not to address in this PR.